dev/tests/static/framework/Magento/TestFramework/CodingStandard/Tool/CopyPasteDetector.php

```diff
@@ -64,35 +74,103 @@ public function canRun()
      * Run tool for files specified
      *
      * @param array $whiteList Files/directories to be inspected
-     * @return int
+     * @return bool
+     */
+    public function run(array $whiteList): bool
+    {
+        $clones = (new Detector(new DefaultStrategy()))->copyPasteDetection(
+            (new Facade())->getFilesAsArray(
+                $whiteList,
+                '',
+                '',
+                $this->getExclude()
+            ),
+            self::MIN_LINES
+        );
+
+        (new PMD($this->reportFile))->processClones($clones);
+        (new Text)->printResult($clones, false);
+
+        return count($clones) == 0;
+    }
+
+    /**
+     * Get exclude params from blacklist
      *
-     * @SuppressWarnings(PHPMD.UnusedLocalVariable)
+     * @return string[]
      */
-    public function run(array $whiteList)
+    private function getExclude(): array
     {
+        $exclude = [];
         $blacklistedDirs = [];
         $blacklistedFileNames = [];
+        $blacklistedPatterns = [];
         foreach ($this->blacklist as $file) {
-            $file = escapeshellarg(trim($file));
+            $file = trim($file);
             if (!$file) {
                 continue;
             }
-            $ext = pathinfo($file, PATHINFO_EXTENSION);
-            if ($ext != '') {
-                $blacklistedFileNames[] = $file;
-            } else {
-                $blacklistedDirs[] = '--exclude ' . $file . ' ';
+            $realPath = realpath(BP . '/' . $file);
+            if ($realPath === false) {
+                $ext = pathinfo($file, PATHINFO_EXTENSION);
+                if ($ext != '') {
+                    $blacklistedFileNames[] = $file;
+                } else {
+                    $blacklistedPatterns[] = $file;
+                }
+                continue;
             }
+
+            $exclude[] = [$realPath];
+            $blacklistedDirs[] = $file;
         }
 
-        $command = $this->getCommand() . ' --log-pmd ' . escapeshellarg($this->reportFile)
-            . ' --names-exclude ' . join(',', $blacklistedFileNames) . ' --min-lines 13 ' . join(' ', $blacklistedDirs)
-            . ' ' . implode(' ', $whiteList);
-        // exec() have to be here since this is test.
-        // phpcs:ignore Magento2.Security.InsecureFunction
-        exec($command, $output, $exitCode);
+        foreach ($blacklistedPatterns as $pattern) {
+            $files = $this->find($pattern, false, $blacklistedDirs);
+            if (empty($files)) {
+                continue;
+            }
+            $exclude[] = $files;
+        }
+
+
+        foreach ($blacklistedFileNames as $fileName) {
+            $files = $this->find($fileName, true, $blacklistedDirs);
+            if (empty($files)) {
+                continue;
+            }
+            $exclude[] = $files;
+        }
+
+        return array_unique(array_merge(...$exclude));
```

Comment:
Technically, there might be empty array in the $exclude, so it's better to replace it with `return array_unique(array_merge([], ...$exclude));`

Reply:
I've tried in interactive shell and it seems that `array_merge(...[])` doesn't throw any warnings and just returns an empty array. Do you see some PHP error while testing?